iree/compiler/Conversion/LinalgToLLVM/ConvImg2ColMatmulConversion.cpp

```diff
@@ -0,0 +1,180 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+class ConvImg2ColMatmulConversion : public OpRewritePattern<linalg::ConvOp> {
+ public:
+  using OpRewritePattern<linalg::ConvOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::ConvOp op,
+                                PatternRewriter &rewriter) const override {
+    auto filterShapeType = op.filter().getType().dyn_cast_or_null<ShapedType>();
+    auto inputShapeType = op.input().getType().dyn_cast_or_null<ShapedType>();
+    auto outputrShapeType =
+        op.output().getType().dyn_cast_or_null<ShapedType>();
+    if (!filterShapeType || !inputShapeType) return failure();
+
+    auto loc = op.getLoc();
+
+    auto inputFeatures =
+        filterShapeType.getShape()[filterShapeType.getRank() - 1];
+    auto outputFeatures = filterShapeType.getShape().back();
+
+    // Col buffer shape (n, d1, d1, d2, ...dn, k1, k2, k3, ...kn, ci)
+    SmallVector<int64_t, 4> colBufferShape;
+    int64_t spatialSize = 1, filterSpatialSize = 1;
+    colBufferShape.push_back(outputrShapeType.getShape()[0]);
+    for (int i = 0; i < op.getNumSpatialDimensions(); ++i) {
+      auto dimSize =
+          outputrShapeType.getShape()[i + op.getNumBatchDimensions()];
+      colBufferShape.push_back(dimSize);
+      spatialSize *= dimSize;
+    }
+    for (int i = 0; i < op.getNumSpatialDimensions(); ++i) {
+      auto dimSize = filterShapeType.getShape()[i];
+      colBufferShape.push_back(dimSize);
+      filterSpatialSize *= dimSize;
+    }
+    colBufferShape.push_back(
+        filterShapeType.getShape()[op.getNumSpatialDimensions()]);
+
+    auto ColBufferMemrefType =
+        MemRefType::get(colBufferShape, filterShapeType.getElementType());
+
+    Value result = rewriter.create<AllocaOp>(loc, ColBufferMemrefType);
+
+    llvm::SmallVector<Value, 4> linalgOpArgs = {op.input(), result};
+
+    // (n, d1, d2, d3, ..., dn, k1, k2, k3, ...kn, ci) ->
+    // (n, d_1 * stride_1 + k_1, d_2 * stride_2 + k_2, ...d_n * stride_n + k_n,
+    // ci)
+    SmallVector<AffineExpr, 4> inputExprs;
+    inputExprs.push_back(rewriter.getAffineDimExpr(0));
+    int spatialDimsOffset = 1;
+    auto kernelDimsOffset = spatialDimsOffset + op.getNumSpatialDimensions();
+    for (int i = 0; i < op.getNumSpatialDimensions(); ++i) {
+      inputExprs.push_back(rewriter.getAffineDimExpr(i + spatialDimsOffset) *
+                               op.getStride(i) +
+                           rewriter.getAffineDimExpr(i + kernelDimsOffset));
+    }
+    inputExprs.push_back(rewriter.getAffineDimExpr(
+        kernelDimsOffset + op.getNumSpatialDimensions()));
+
+    auto nloops = colBufferShape.size();
```

Comment:
This seems like it is skipping a lot of steps at once directly going to Affine tiled loops? Can we rephrase this problem as a tiling + promotion problem. I am fairly convinced that it is possible to do it that way. Such a one off pattern will make it hard to think about fusion of operations, etc.

So overall, can we phrase this question as a Tiling + Promotion problem. I can help you write the patterns that will get you what you want, or at least we will know if this is not a tiling + promotion problem.

Reply:
This actually assumes `linalg.conv` already tiled otherwise the column buffer will be massive (I should add an optional size limit). The pattern here to rewrite the `linalg.conv` itself